packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -100,29 +120,120 @@ export class SerializedStateManager {
 		supportGetSnapshotApi: boolean,
 	) {
 		if (this.pendingLocalState === undefined) {
-			const { snapshot, version } = supportGetSnapshotApi
-				? await fetchISnapshot(this.mc, this.storageAdapter, specifiedVersion)
-				: await fetchISnapshotTree(this.mc, this.storageAdapter, specifiedVersion);
-			const baseSnapshot: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
-				? snapshot.snapshotTree
-				: snapshot;
-			assert(baseSnapshot !== undefined, 0x8e4 /* Snapshot should exist */);
+			const { snapshotTree, version } = await getSnapshotTree(
+				this.mc,
+				this.storageAdapter,
+				supportGetSnapshotApi,
+				specifiedVersion,
+			);
 			// non-interactive clients will not have any pending state we want to save
 			if (this.offlineLoadEnabled) {
 				const snapshotBlobs = await getBlobContentsFromTree(
-					baseSnapshot,
+					snapshotTree,
 					this.storageAdapter,
 				);
-				this.snapshot = { baseSnapshot, snapshotBlobs };
+				this.snapshot = { baseSnapshot: snapshotTree, snapshotBlobs };
 			}
-			return { baseSnapshot, version };
+			return { baseSnapshot: snapshotTree, version };
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi)
+				.catch((error) => {
+					this.mc.logger.sendErrorEvent(
```

Comment:
A lot of logging work is still needed. Ignore this

Reply:
wrapping the body of the refreshSnapshot in  a performance event would be a better pattern